### PR TITLE
[WIP] Add options call for network routers

### PR DESCRIPTION
**Pull request by GilbertCherrie**
Depends on:  https://github.com/ManageIQ/manageiq-providers-openstack/pull/731

Added an options call for network routers.

Calling example with no cloud_network_id:
OPTIONS "http://admin:smartvm@localhost:3000/api/network_routers?ems_id=8"
`{"attributes":["cidr","cloud_tenant_id","description","ems_id","ems_ref","enabled","external_facing","extra_attributes","id","name","orchestration_stack_id","provider_network_type","provider_physical_network","provider_segmentation_id","resource_group_id","shared","status","type","vlan_transparent"],"virtual_attributes":["href_slug","maximum_transmission_unit","port_security_enabled","qos_policy_id","region_description","region_number","total_vms"],"relationships":["cloud_subnets","cloud_tenant","custom_action_buttons","custom_actions","custom_button_events","custom_button_sets","custom_buttons","ext_management_system","floating_ips","network_ports","network_routers","orchestration_stack","private_networks","public_network_routers","public_network_vms","public_networks","resource_group","security_groups","taggings","tags","vms"],"subcollections":["tags"],"data":{"form_schema":{"fields":[{"component":"select","id":"cloud_tenant_id","name":"cloud_tenant_id","key":"id-8","label":"Cloud Tenant Placement","placeholder":"\u003cChoose\u003e","validateOnMount":true,"validate":[{"type":"required","message":"Required"}],"isRequired":true,"options":[{"label":"cloud-user-demo","value":"1"},{"label":"admin","value":"2"},{"label":"openshift_demo","value":"3"},{"label":"testtel","value":"4"},{"label":"cloud-southeast","value":"5"},{"label":"Loic Tenant","value":"6"},{"label":"cloudwest","value":"7"},{"label":"Massachusetts","value":"8"},{"label":"testetot","value":"10"},{"label":"test-ivy","value":"13"},{"label":"test-ivr","value":"12"}],"includeEmpty":true,"clearOnUnmount":true},{"component":"sub-form","title":"Router Information","id":"routerInformation","name":"routerInformation","fields":[{"component":"text-field","id":"router_name","name":"name","label":"Router Name","validateOnMount":true,"validate":[{"type":"required","message":"Required"}],"isRequired":true,"clearOnUnmount":true},{"component":"switch","id":"admin_state_up","name":"admin_state_up","label":"Administrative State","onText":"Up","offText":"Down","initialValue":true}]},{"component":"sub-form","title":"External Gateway","id":"externalGateway","name":"externalGateway","fields":[{"component":"switch","id":"enable","name":"enable","label":"Enable","onText":"Yes","offText":"No"},{"component":"switch","id":"source_nat","name":"source_nat","label":"Source NAT","onText":"Yes","offText":"No","condition":{"when":"enable","is":true},"initialValue":true},{"component":"select","id":"network","name":"network","key":"id-8","label":"Network","placeholder":"\u003cChoose\u003e","includeEmpty":true,"clearOnUnmount":true,"condition":{"when":"enable","is":true},"options":[{"label":"ext","value":"6b3d0c3b-8b68-4c26-a493-5be44d160241,50"}]}]}]}}}%`

Can also call with a cloud_network_id in order to return only fields that are allowed to be edited while the rest are disabled.
OPTIONS "http://admin:smartvm@localhost:3000/api/network_routers/${cloud_network_id}

@miq-bot add-label enhancement
@miq-bot add-label dependencies
@miq-bot add_reviewer @Fryguy

### CODE REVIEW

File: spec/requests/network_routers_spec.rb [hunk 1]
```diff
@@ -204,4 +204,30 @@
       expect(response.parsed_body).to include(expected)
     end
   end
+
+  describe 'OPTIONS /api/network_routers' do
```
Comment (bdunne) [inline spec/requests/network_routers_spec.rb:208]: This should probably mention that it's testing OPTIONS with `ems_id=...`

Comment (bdunne) [reply spec/requests/network_routers_spec.rb:208]: Can you also add a test for OPTIONS without `ems_id=`?

File: spec/requests/network_routers_spec.rb [hunk 2]
```diff
@@ -204,4 +204,30 @@
...
+    it 'returns a DDF schema for add when available via OPTIONS' do
+      zone = FactoryBot.create(:zone, :name => "api_zone")
```
Comment (bdunne) [inline spec/requests/network_routers_spec.rb:210]: The zone doesn't seem important to your test, the `ems_network` factory should already take care of it for you.

File: app/controllers/api/network_routers_controller.rb [hunk 3]
```diff
@@ -27,13 +27,40 @@ def delete_resource(type, id, _data = {})
...
+
         task_id = network_router.delete_network_router_queue(User.current_user.id)
         action_result(true, "Deleting #{network_router_ident(network_router)}", :task_id => task_id)
       end
     end
 
+    def options
+      if params.key?("ems_id")
+        options_by_ems_id
+      elsif params.key?("id")
```
Comment (bdunne) [inline app/controllers/api/network_routers_controller.rb:39]: Should this be a higher priority than `ems_id`?

File: spec/requests/network_routers_spec.rb [hunk 4]
```diff
@@ -204,4 +204,36 @@
...
+      expect(response).to have_http_status(:ok)
+    end
+  end
+
+  describe 'OPTIONS /api/network_routers/:id' do
+    it 'returns a DDF schema for edit when available via OPTIONS' do
+      network_router = FactoryBot.create(:network_routers)
+      allow(NetworkRouter).to receive(:find).with(network_router.id.to_s).and_return(network_router)
+      allow(Rbac).to receive(:filtered_object).and_return(network_router)
+      expect(network_router).to receive(:params_for_edit).and_return('foo')
```
Comment (Fryguy) [inline spec/requests/network_routers_spec.rb:233]: elsewhere I believe we called this params_for_update.  @agrare Can you confirm?

Comment (agrare) [reply spec/requests/network_routers_spec.rb:233]: That's correct CloudVolumes calls this `params_for_update`

Comment (GilbertCherrie) [reply spec/requests/network_routers_spec.rb:233]: I can change it to params_for_update

**Comment by kbrock** [comment]
I really like the implementation [[here]](https://github.com/manageiq/manageiq-api/blob/master/app/controllers/api/cloud_volumes_controller.rb#L51-L65)

since type is dynamic and most of these are based upon an ems or an id, it feels like this could be generic. At least for the cloud volume.

@agrare can this be extracted out?

**Comment by agrare** [comment]
Yeah that could definitely be extracted to a mixin

**Comment by GilbertCherrie** [comment]
@agrare @kbrock Is this PR still valid or should I change it? Not sure how to implement a mixin.

**Comment by agrare** [comment]
@GilbertCherrie you're good to keep going this way we'll tackle any refactoring in a separate PR

**Comment by kbrock** [comment]
closing and deferring to #1079 
This was the base for that PR and the other adds a bunch of tests and hopefully establishes a pattern we can use for other models.

**Pull request closed without merging**